### ISSUE TITLE
Add support for DropdownV2 to have categories

### ISSUE THREAD

**Issue by meganwalker-ibm** [carbon-design-system/carbon#2595]
### Summary

In Dropdown we managed to get categories of items rendered like so, by using fake items in the data structure, and ignoring the click events etc.

<img width="277" alt="screen shot 2018-10-01 at 10 16 22" src="https://user-images.githubusercontent.com/20321922/46280168-146e3080-c563-11e8-8c07-d470f3fed1a3.png">

(This example is from existing public functionality and shows a list of places a slack message could be sent. After the end of the list of channels, you see the category heading 'USERS' followed by a list of selectable user options).

In DropdownV2 we have to pass in just an array of plain objects that, as far as I can see, have no way of specifying categories. We need to be able to specify categories to migrate to DropdownV2. Ideally we don't want to have to hack them in either, but instead have first class support like the `Select` component does.

### Justification

Dropdown is disappearing, and being replaced with DropdownV2. This would also provide feature parity with the `Select` carbon component.

### Desired UX and success metrics

Items should be able to be grouped under unselectable categories. This provides structure to what could be a long list of otherwise indistinguishable options. Ideally it should look similar to what we have for our existing `Dropdown` usage.

### "Must have" functionality
- Ability to provide category information in some format.
- Categories should not be selectable
- Categories should be rendered differently to provide differentiation.

### Specific timeline issues / requests

Before Dropdown disappears.

### Available extra resources

It's unclear if we'll have the time or resource to implement such functionality ourselves - even if we can reach an agreement here for how categories should be injected and styling should look.

**Comment by lovemecomputer** [comment]
@meganwalker-ibm cool yeah, this functionality would be rad! like @joshblack said yeah design will weigh in on this, and determine if we would just do this implementation or if it could fit in the context of other patterns :) but if y'all end up with a new component/option to commit, we'll definitely look at that too!

**Comment by aagonzales** [comment]
Yes, I agree with @designertyler it needs a little polish. Except I hesitate to use semi-bold for categories since it could be mistaken for a selected state because of its emphasis. 

![image](https://user-images.githubusercontent.com/11670886/46437350-c868ea80-c720-11e8-8445-ecad9c1e64cd.png)

Or if they don't need labels maybe just dividers to show association, like on os menu. Just some ideas

![image](https://user-images.githubusercontent.com/11670886/46437322-b8e9a180-c720-11e8-91f9-b1d0ba8d4807.png)

**Comment by ashharrison90** [comment]
Hi, can someone reopen this issue? I'm about to start work on it.
In terms of implementation, my initial thought is something like passing a `categoryKey` to `Dropdown` 

So say your items look something like:
```
const items = [
  {
    id: 'germany',
    text: 'Germany',
    continent: 'Europe',
  },
  {
    id: 'uk',
    text: 'United Kingdom',
    continent: 'Europe',
  },
  {
    id: 'china',
    text: 'China',
    continent: 'Asia',
  },
  {
    id: 'japan',
    text: 'Japan',
    continent: 'Asia',
  },
];
```
Then you'd have:
```
<Dropdown
  ...normalProps
  categoryKey='continent'
\>
```
And `Dropdown` would then group options under the category headers of `Asia` and `Europe`.

How does this sound?

**Comment by ashharrison90** [comment]
which ends up looking something like this:
![image](https://user-images.githubusercontent.com/20999846/67418031-f5a16880-f5c1-11e9-82c5-ac6d6e26fe99.png)

**Comment by asudoh** [comment]
@carbon-design-system/design Are we still interested in this feature? Thanks!

**Comment by tw15egan** [comment]
@joshblack  this still something we plan on supporting? Looking at https://github.com/carbon-design-system/carbon/pull/4449 and trying to see what the next step is for this feature

**Comment by joshblack** [comment]
@tw15egan I think we definitely want to, but IIRC there were some issues with a11y that @dakahn flagged. I think on our end we wanted to fix up the a11y issues in these components first and then move forward with category support emulating: https://dequelabs.github.io/combobo/demo/ but I'm not sure if there was a hiccup with that and screen readers, or not, so will defer to DA 🤔

**Comment by dakahn** [comment]
We need to figure out which dropdown menu style components should have category support and which should not. Category support will be tricky to get right from an accessibility perspective (functioning cross browser, cross screen reader, cross platform). 

Actually @joshblack last I remember us discussing this we did not want to move forward with categories in Dropdown and wanted to point interested users toward another component? @abbeyhrt pinging you in here since I remember you being part of this discussion

**Comment by joshblack** [comment]
@dakahn if I understand correctly, I think most variants could need category support, including:

- Listbox (dropdown)
  - Single select
  - Multi select
- Combobox
  - Single select
  - Multi select

Not sure about other components, I think we'd need to include them here unless we're unable to make them work accessibly 🤔

**Comment by designertyler** [comment]
Next steps for design is to define rules for usage and final spec. Maybe look at existing components with category labels.
